src/HttpServer.gren

```diff
@@ -0,0 +1,453 @@
+                        }
+        
+            GotRequest req res ->
+                let
+                    body = "You requested: " ++ Url.toString req.url
+                in
+                { model = model
+                , command =
+                    res
+                        |> Response.setHeader "Content-type" "text/html"
+                        |> Response.setBody ("<html>" ++ body ++ "</html>")
+                        |> Response.send
+                }
+
+    subscriptions : Model -> Sub Msg
+    subscriptions model =
+        case model.server of
+            Just server ->
+                Http.onRequest server GotRequest
+
+            Nothing ->
+                Sub.none
+
+See [examples/http-server](https://github.com/gren-lang/node/tree/main/examples/http-server) for more.
+
+## Initialization
+
+@docs Permission, Server, ServerError, initialize, createServer
+
+## Requests
+
+@docs Request, Method, bodyAsString, bodyFromJson, toRequest, requestInfo
+
+## Responding to requests
+
+@docs onRequest
+
+See [HttpServer.Response](HttpServer.Response) for more details on responding to requests.
+
+-}
+
+import Bytes exposing (Bytes)
+import Bytes.Decode as Decode
+import Dict exposing (Dict)
+import Init
+import Internal.Init
+import Json.Encode
+import Json.Decode
+import Node
+import Task exposing (Task)
+import Gren.Kernel.HttpServer
+import HttpServer.Response exposing (Response(..))
+import Url exposing (Url, Protocol(..))
+
+
+-- INITIALIZATION
+
+
+{-| The permission to start a [`Server`](HttpServer.Server).
+
+You get this from [`initialize`](HttpServer.initialize).
+-}
+type Permission
+    = Permission
+
+{-| The HTTP server.
+-}
+type Server
+    -- Note: Actual implementation in Kernel code
+    = Server
+
+
+{-| Error code and message from node.
+Most likely from a failed attempt to start the server (e.g. `EADDRINUSE`).
+Refer to the [node docs](https://nodejs.org/docs/latest-v18.x/api/errors.html) for details.
+-}
+type ServerError =
+    ServerError String String
+
+
+{-| Initialize the [`HttpServer`](HttpServer) module and get permission to create a server.
+-}
+initialize : Init.Task Permission
+initialize =
+    Task.succeed Permission
+        |> Internal.Init.Task
+
+
+
+{-| Task to initialize a [`Server`](HttpServer#Server).
+-}
+createServer : Permission -> { host : String, port_ : Int } -> Task ServerError Server
+createServer _ options =
+    Gren.Kernel.HttpServer.createServer options.host options.port_
+
+
+-- REQUESTS
+
+
+{-| An incoming HTTP reqest.
+-}
+type alias Request =
+    { headers : Dict String String
+    , method : Method
+    , body : Bytes
+    , url : Url
+    }
+
+
+{-| HTTP request methods.
+-}
+type Method
+    = GET
+    | HEAD
+    | POST
+    | PUT
+    | DELETE
+    | CONNECT
+    | TRACE
+    | PATCH
+    | UNKNOWN
+
+
+{-| Turn the pieces of a request into a [`Request`](HttpServer.Request) record.
+
+This is used internally, you shouldn't need it in your own code.
```

Comment:
Does it need to be exposed? I believe kernel code ignores exposed/not-exposed settings

Reply:
Very glad to discover it doesn't need to be exposed for the kernel!